pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -348,10 +348,68 @@ func makeFirewallName(name string) string {
 	return fmt.Sprintf("k8s-fw-%s", name)
 }
 
+func (gce *GCECloud) getAddress(name, region string) (string, bool, error) {
+	address, err := gce.service.Addresses.Get(gce.projectID, region, name).Do()
+	if err == nil {
+		return address.Address, true, nil
+	}
+	if isHTTPErrorCode(err, http.StatusNotFound) {
+		return "", false, nil
+	}
+	return "", false, err
+}
+
+func ownsAddress(ip net.IP, addrs []*compute.Address) bool {
+	ipStr := ip.String()
+	for _, addr := range addrs {
+		if addr.Address == ipStr {
+			return true
+		}
+	}
+	return false
+}
+
 // EnsureTCPLoadBalancer is an implementation of TCPLoadBalancer.EnsureTCPLoadBalancer.
 // TODO(a-robinson): Don't just ignore specified IP addresses. Check if they're
 // owned by the project and available to be used, and use them if they are.
 func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, externalIP net.IP, ports []*api.ServicePort, hosts []string, affinityType api.ServiceAffinity) (*api.LoadBalancerStatus, error) {
+	if externalIP == nil {
+		glog.V(2).Info("Checking if the external ip address already exists: %s", name)
+		address, exists, err := gce.getAddress(name, region)
+		if err != nil {
+			return nil, fmt.Errorf("error looking for gce address: %v", err)
+		}
+		if !exists {
+			// Note, though static addresses that _aren't_ in use cost money, ones that _are_ in use don't.
+			// However, quota is limited.  TODO: determine quota here
```

Comment:
I just tested it.  "creating" an IP with a specified IP upgrades it to
static and "deleteing" an in-use static IP downgrades it to ephemeral.
This is a great hack!

Who can get on this?  I think we want it for 1.1, but it needs to get some
soak ASAP.  Or I guess it come come in 1.1.1

On Thu, Oct 8, 2015 at 8:15 PM, Tim Hockin <thockin@google.com> wrote:

> Whoah, I did not know you could do that!  Before I said that out loud, I
> checked - and you CAN.
>
> That sounds like a really good hack.  I'm trying to test it out right now,
> but I can't seem to get GCE to make a load-balancer (sigh).
>
> On Thu, Oct 8, 2015 at 5:24 PM, CJ Cullen <notifications@github.com>
> wrote:
>
>> In pkg/cloudprovider/providers/gce/gce.go
>> <https://github.com/kubernetes/kubernetes/pull/13335#discussion_r41586467>
>> :
>>
>> >  // EnsureTCPLoadBalancer is an implementation of TCPLoadBalancer.EnsureTCPLoadBalancer.
>> >  // TODO(a-robinson): Don't just ignore specified IP addresses. Check if they're
>> >  // owned by the project and available to be used, and use them if they are.
>> >  func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, externalIP net.IP, ports []*api.ServicePort, hosts []string, affinityType api.ServiceAffinity) (*api.LoadBalancerStatus, error) {
>> > +	if externalIP == nil {
>> > +		glog.V(2).Info("Checking if the external ip address already exists: %s", name)
>> > +		address, exists, err := gce.getAddress(name, region)
>> > +		if err != nil {
>> > +			return nil, fmt.Errorf("error looking for gce address: %v", err)
>> > +		}
>> > +		if !exists {
>> > +			// Note, though static addresses that _aren't_ in use cost money, ones that _are_ in use don't.
>> > +			// However, quota is limited.  TODO: determine quota here
>>
>> Do we really need to hold on to static IPs indefinitely? We could get
>> around quota problems by only creating a static IP for the life of the
>> forwarding rule update.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/13335/files#r41586467>.
>>
>
>